modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/reducer.ts

```diff
@@ -22,7 +22,18 @@ const STATES = {
 
 export {STATES};
 
-export default function reducer(state, action) {
+interface State {
+	current: {show: boolean};
+	nextTarget?: HTMLElement;
+	target?: HTMLElement;
+	timestamp?: number;
+}
+
+type Action =
+	| {type: 'show'; target?: HTMLElement}
+	| {type: 'hide' | 'hideDelayCompleted' | 'showDelayCompleted'};
```

Comment:
Probably clearer as:

```ts
type Action =
	| {type: 'hide'}
	| {type: 'hideDelayCompleted'}
	| {type: 'showDelayCompleted'}
	| {type: 'show'; target?: HTMLElement};
```

Given that types are elided from the output, both versions are 0 bytes anyway, so may as well go with the most readable one. `#microoptimization`

Also, sorting, because `#ThisIsLiferay`.

![this-is-sparta](https://user-images.githubusercontent.com/7074/115835593-bb16c700-a416-11eb-961d-72fb5e7e9530.gif)

Backup gif:

![this-is-the-way](https://user-images.githubusercontent.com/7074/115835709-de417680-a416-11eb-893c-bd6126f1a16f.gif)